Order History partially paid fixed `READY_FOR_REVIEW`

jpgcz opened the pull request:
# Order History partially paid fixed PR `READY_FOR_REVIEW`

## Summary Of Changes :fire:
The "Замовлення Частково оплачено" event didn't appear in the "Історія замовлення" order work log after the payment was added by the administrator and the actual amount of the order was greater than the paid amount.

## Issue Link 🔗 
[Git Issue Card #4507](https://github.com/ita-social-projects/GreenCity/issues/4507)

## Added
- Added a validation in buildPaymentEntity on the Service Implement, to check if the order is fully or partially paid
- Added the ORDER_HALF_PAID description

## How to test
This could be tested with Swagger also
1. First you need to run GreenCityClient
2. Create an order and partially paid it
3. The status of the order should be "Partially paid" | "Замовлення Oплачено" in Order History
4. That's it! ✔️

# PR Checklist Forms ✔️ 
- [x] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [x] JIRA/ GitHub Issue number & title in PR title (ISSUE-XXXX: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [x] `NEED_REVIEW` and `READY_FOR_REVIEW` labels are added.
- [x]  All files reviewed before sending to reviewers

jpgcz committed:
Order History partially paid fixed

jpgcz committed:
Order History partially paid fixed

jpgcz committed:
Order History partially paid fixed

jpgcz committed:
Merge branch 'dev' of https://github.com/ita-social-projects/GreenCityUBS into bug_#4507_UBS_Admin_History_Partially_Paid

jpgcz committed:
Test fixed

jpgcz committed:
Merge branch 'dev' of https://github.com/ita-social-projects/GreenCityUBS into bug_#4507_UBS_Admin_History_Partially_Paid

# Conflicts:
#	service/src/test/java/greencity/service/ubs/UBSManagementServiceImplTest.java

jpgcz committed:
Test added and fixed format

jpgcz committed:
Test added and fixed format

RomaMocherniuk commented:
Please increase test coverage at least to 95%
![image](https://user-images.githubusercontent.com/15122465/226858770-b5103feb-dc69-4daf-be87-0d550bb18cd2.png)

RomaMocherniuk commented on service-api/src/main/java/greencity/constant/OrderHistory.java:
```diff
@@ -6,6 +6,7 @@ public final class OrderHistory {
     public static final String ORDER_BROUGHT_IT_HIMSELF = "Статус Замовлення - Привезе сам";
     public static final String ORDER_CANCELLED = "Статус Замовлення - Скасовано";
     public static final String ORDER_PAID = "Замовлення Оплачено";
+    public static final String ORDER_HALF_PAID = "Замовлення Частково оплачено";
```
The message says: `"Order is partly paid"` and the variable name is `ORDER_HALF_PAID`
Please change the message or the variable name to have consistency between them

RomaMocherniuk commented on service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java:
```diff
@@ -1379,7 +1379,13 @@ private Payment buildPaymentEntity(Order order, ManualPaymentRequestDto paymentR
             employee.getFirstName() + "  " + employee.getLastName(), order);
-        eventService.save(OrderHistory.ORDER_PAID, OrderHistory.SYSTEM, order);
+        if (order.getOrderPaymentStatus() != null) {
+            if (order.getOrderPaymentStatus().equals(OrderPaymentStatus.PAID)) {
```
we can use `== `for enums
```
order.getOrderPaymentStatus() == OrderPaymentStatus.PAID
```

RomaMocherniuk commented on service/src/test/java/greencity/service/ubs/UBSManagementServiceImplTest.java:
```diff
@@ -348,6 +320,67 @@ void checkUpdateManualPayment() {
+        when(employeeRepository.findByEmail("test@gmail.com")).thenReturn(Optional.of(employee));
+        when(tariffsInfoRepository.findTariffsInfoByIdForEmployee(anyLong(), anyLong()))
+            .thenReturn(Optional.of(tariffsInfo));
+        when(orderRepository.getOrderDetails(1L)).thenReturn(Optional.of(order));
```
please add `verify` for it

RomaMocherniuk commented on service/src/test/java/greencity/service/ubs/UBSManagementServiceImplTest.java:
```diff
@@ -348,6 +320,67 @@ void checkUpdateManualPayment() {
+            .settlementdate("02-08-2021").amount(500L).receiptLink("link").paymentId("1").build();
+        Employee employee = ModelUtils.getEmployee();
+        when(orderRepository.findById(1L)).thenReturn(Optional.of(order));
+        when(employeeRepository.findByEmail("test@gmail.com")).thenReturn(Optional.of(employee));
```
please add `verify` for it

RomaMocherniuk commented on service/src/test/java/greencity/service/ubs/UBSManagementServiceImplTest.java:
```diff
@@ -348,6 +320,67 @@ void checkUpdateManualPayment() {
+        when(employeeRepository.findByEmail("test@gmail.com")).thenReturn(Optional.of(employee));
+        when(tariffsInfoRepository.findTariffsInfoByIdForEmployee(anyLong(), anyLong()))
+            .thenReturn(Optional.of(tariffsInfo));
+        when(orderRepository.getOrderDetails(1L)).thenReturn(Optional.of(order));
```
please add `verify`

RomaMocherniuk commented on service/src/test/java/greencity/service/ubs/UBSManagementServiceImplTest.java:
```diff
@@ -348,6 +320,67 @@ void checkUpdateManualPayment() {
+            .settlementdate("02-08-2021").amount(200L).receiptLink("link").paymentId("1").build();
+        Employee employee = ModelUtils.getEmployee();
+        when(orderRepository.findById(1L)).thenReturn(Optional.of(order));
+        when(employeeRepository.findByEmail("test@gmail.com")).thenReturn(Optional.of(employee));
```
please add `verify`

jpgcz replied:
can you please specify what do I need to verify? All the "verify" are at the end, or there's something unverified? ty

RomaMocherniuk replied:
In this test you have the next mocks:
```
when(orderRepository.findById(1L)).thenReturn(Optional.of(order));
when(employeeRepository.findByEmail("test@gmail.com")).thenReturn(Optional.of(employee));
when(tariffsInfoRepository.findTariffsInfoByIdForEmployee(anyLong(), anyLong())).thenReturn(Optional.of(tariffsInfo));
when(orderRepository.getOrderDetails(1L)).thenReturn(Optional.of(order));
when(paymentRepository.save(any())).thenReturn(payment);
doNothing().when(eventService).save(OrderHistory.ADD_PAYMENT_MANUALLY + 1, "Петро" + "  " + "Петренко", order);

```
You added `verify` for:
```
verify(eventService, times(1)).save("Замовлення Частково оплачено", "Система", order);
verify(paymentRepository, times(1)).save(any());
verify(orderRepository, times(1)).findById(1L);
verify(tariffsInfoRepository).findTariffsInfoByIdForEmployee(anyLong(), anyLong());
```
I can't see `verify` for:
```
when(employeeRepository.findByEmail("test@gmail.com")).thenReturn(Optional.of(employee));
when(orderRepository.getOrderDetails(1L)).thenReturn(Optional.of(order));
```
Could you please add them?
Thank you

RomaMocherniuk commented on service/src/test/java/greencity/service/ubs/UBSManagementServiceImplTest.java:
```diff
@@ -348,6 +320,67 @@ void checkUpdateManualPayment() {
+        when(orderRepository.getOrderDetails(1L)).thenReturn(Optional.of(order));
+        when(paymentRepository.save(any()))
+            .thenReturn(payment);
+        doNothing().when(eventService).save(OrderHistory.ADD_PAYMENT_MANUALLY + 1, "Петро" + "  " + "Петренко", order);
```
please add `verify`

RomaMocherniuk commented on service/src/test/java/greencity/service/ubs/UBSManagementServiceImplTest.java:
```diff
@@ -348,6 +320,67 @@ void checkUpdateManualPayment() {
+        when(paymentRepository.save(any()))
+            .thenReturn(payment);
+        doNothing().when(eventService).save(OrderHistory.ADD_PAYMENT_MANUALLY + 1, "Петро" + "  " + "Петренко", order);
+        ubsManagementService.saveNewManualPayment(1L, paymentDetails, null, "test@gmail.com");
```
`ubsManagementService.saveNewManualPayment` returns `ManualPaymentResponseDto`
Could you please check if it contains the expected data?

RomaMocherniuk commented on service/src/test/java/greencity/service/ubs/UBSManagementServiceImplTest.java:
```diff
@@ -348,6 +320,67 @@ void checkUpdateManualPayment() {
+        when(paymentRepository.save(any()))
+            .thenReturn(payment);
+        doNothing().when(eventService).save(OrderHistory.ADD_PAYMENT_MANUALLY + 1, "Петро" + "  " + "Петренко", order);
+        ubsManagementService.saveNewManualPayment(1L, paymentDetails, null, "test@gmail.com");
```
same here
`ubsManagementService.saveNewManualPayment` returns `ManualPaymentResponseDto`
Could you please check if it contains the expected data?

RomaMocherniuk reviewed (changes requested):
Left some comments

RomaMocherniuk commented on service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java:
```diff
@@ -1379,7 +1379,13 @@ private Payment buildPaymentEntity(Order order, ManualPaymentRequestDto paymentR
         eventService.save(OrderHistory.ADD_PAYMENT_MANUALLY + paymentRequestDto.getPaymentId(),
             employee.getFirstName() + "  " + employee.getLastName(), order);
-        eventService.save(OrderHistory.ORDER_PAID, OrderHistory.SYSTEM, order);
+        if (order.getOrderPaymentStatus() != null) {
```
I see in this method we have that logic
```
 Payment payment = Payment.builder()
            .settlementDate(LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE))
            .amount(paymentRequestDto.getAmount())
            .paymentStatus(PaymentStatus.PAID)
           ...
```

Should we check `order.getOrderPaymentStatus()` before set `paymentStatus`?

jpgcz replied:
Yes, I made that verification now

RomaMocherniuk commented on service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java:
```diff
@@ -1379,7 +1379,13 @@ private Payment buildPaymentEntity(Order order, ManualPaymentRequestDto paymentR
             .orElseThrow(() -> new EntityNotFoundException(EMPLOYEE_NOT_FOUND));
         eventService.save(OrderHistory.ADD_PAYMENT_MANUALLY + paymentRequestDto.getPaymentId(),
             employee.getFirstName() + "  " + employee.getLastName(), order);
-        eventService.save(OrderHistory.ORDER_PAID, OrderHistory.SYSTEM, order);
```
I see we call 
```
 eventService.save(OrderHistory.ADD_PAYMENT_MANUALLY + paymentRequestDto.getPaymentId(),
            employee.getFirstName() + "  " + employee.getLastName(), order);
```
And after that, we check if `(order.getOrderPaymentStatus() != null)` and if it is not `null` we call 
```
    eventService.save(OrderHistory.ORDER_PAID, OrderHistory.SYSTEM, order);
```
or
```
eventService.save(OrderHistory.ORDER_HALF_PAID, OrderHistory.SYSTEM, order);
```
Based on this logic we could have a situation when payment is `ADD_PAYMENT_MANUALLY` but not `ORDER_PAID` or `ORDER_HALF_PAID`
Is this case valid?

jpgcz replied:
I'll ask viktoriya about it, I think that we may need to verify also for unpaid orders, but I'll check

jpgcz commented on service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java:
```diff
@@ -1379,7 +1379,13 @@ private Payment buildPaymentEntity(Order order, ManualPaymentRequestDto paymentR
             .orElseThrow(() -> new EntityNotFoundException(EMPLOYEE_NOT_FOUND));
         eventService.save(OrderHistory.ADD_PAYMENT_MANUALLY + paymentRequestDto.getPaymentId(),
             employee.getFirstName() + "  " + employee.getLastName(), order);
-        eventService.save(OrderHistory.ORDER_PAID, OrderHistory.SYSTEM, order);
```
There are four Payment Statuses: Paid, Half Paid, Unpaid, Payment Refunded.
For the payment entity, we don't need to have cover for all events, because those other 2 events are covered in other methods, like when you cancel an order. In case we don't pay o half pay the order, I don't think that it should display an event for that

jpgcz committed:
Added method to verify status order and changed the displayed message

Markiro1 committed:
- changed test 1+ condition

Markiro1 committed:
Merge branch 'dev' into bug_#4507_UBS_Admin_History_Partially_Paid

Markiro1 committed:
- added some method in models utils and test in UBSManagementServiceImplTest

Markiro1 committed:
- formatted code

Markiro1 committed:
- delete 1 check in method

Markiro1 committed:
- rechange check

Markiro1 committed:
- removed test methods and removed unnecessary check

Markiro1 committed:
- added test when order is null

Markiro1 committed:
- redundant test removed
- unnecessary order checks have been removed, since before calling the private method there is already a check to see if the order is empty

Markiro1 committed:
- fix 1 bug

Markiro1 committed:
- fix 1 bug

Markiro1 committed:
- added getOrderStatusTranslations() for test orderStatusTranslationRepository.findAllBy()
- added new verify in getOrderStatusDataTest

Markiro1 committed:
- added new verify
- added new method getOrderStatusPaymentTranslations() in ModelUtils

Markiro1 committed:
- added new test

Markiro1 committed:
- added new test
- change getOrdersStatusDoneDto()

Markiro1 committed:
- change test

Markiro1 committed:
- change test
- added new test

Markiro1 committed:
- formatted code

Markiro1 committed:
- change test

Markiro1 committed:
- change imports

Markiro1 committed:
Merge branch 'dev' into bug_#4507_UBS_Admin_History_Partially_Paid
